### PR TITLE
docs: Document eBPF IP preservation solution

### LINKED ISSUE #742: docs: add hint to calicos eBPF dataplane for IP preservation
Calico has added a new and exciting feature that creates a eBPF dataplane replacing kube-proxy. 
This enables the preservation of the source IP and other things.

One should add a note or even dedicate a section in the docs to this feature, since it's a viable solution for the souce IP problem, should one use calico.

### PR DESCRIPTION
**Pull request by onmyflow**
This documentation highlights the implementation of the eBPF dataplane
in CNIs, resulting in successful packet source IP preservation and improvements
to network efficiencies.

Fixed #742

Signed-off-by: Daniel Schneider <daniel.schneider@eramux.com>

### CODE REVIEW

**Comment by onmyflow** [comment]
Friendly ping: @daxmc99 @johananl @rata

File: website/content/usage/_index.md
Comment (daxmc99) [inline website/content/usage/_index.md:171]:
```suggestion
Another CNI that supports the eBPF dataplane is Cilium. You can read up more
```

**Review by daxmc99 — APPROVED** [review]
Looks good! I'll wait for another maintainer to approve before merging though

**Comment by champtar** [comment]
@onmyflow thanks a lot for writing docs, this is as important as code !!
IP source preservation is not specific to eBPF, it's just a consequence of DSR (direct server return), that you can have without eBPF, example https://github.com/cloudnativelabs/kube-router/blob/master/docs/user-guide.md#direct-server-return

**Comment by onmyflow** [comment]
@champtar Interesting! I'm actually not all too versed in network specifics like those, that's good to know. 
So how should I proceed then?
* Should I mention that DSR is a foundation for eBPF, which is a viable solution to IP preservation? 
* Should I rewrite the whole section to only mention DSR, which is the correct solution but might be a bit technical for your everyday DevOps engineer?

At least for me, I wouldn't even know where to search for a solution to "MetalLB IP preservation". So either way it will save others quite some time when faced with the same problem.

**Comment by champtar** [comment]
You can have DSR without eBPF, you can have IP source preservation without DSR (weird but technically feasible), but DSR imply IP source preservation as you are responding directly you need the source IP/port.
https://www.loadbalancer.org/blog/layer-4-vs-layer-7-load-balancing-we-still-love-dsr/

**Review by m-yosefpor — CHANGES_REQUESTED** [review]

File: website/content/usage/_index.md
Comment (joestringer) [inline website/content/usage/_index.md:172]: Hi, Cilium developer here. :wave: 

The link here is pointing to the eBPF and XDP reference guide, which is a useful reference for eBPF but does not describe the Cilium implementation and usage of eBPF. Is there any reason not to just link to https://cilium.io here?

Also, for what it's worth I wouldn't necessarily describe Cilium as "Another CNI that supports the eBPF dataplane" since Cilium doesn't have another mode; Cilium was built from the ground up to leverage eBPF as we described in the blog post here: https://cilium.io/blog/2018/04/24/cilium-security-for-age-of-microservices . Given that this documentation is for a third party project, I think it probably makes sense to describe these external projects using the taglines on the front pages of the external project's website, so for instance for Cilium "Cilium provides eBPF-based Networking, Observability and Security". This is probably the more neutral way to describe the projects in the way that the projects themselves present themselves. For readers that are interested in more details, they can then follow the links and perform additional research on the project.

If you do want a more specific link to documentation regarding eBPF and client source IP preservation, the appropriate link is https://docs.cilium.io/en/stable/gettingstarted/kubeproxy-free/#client-source-ip-preservation.

File: website/content/usage/_index.md
Comment (tgraf) [inline website/content/usage/_index.md]: Thanks for thinking of us!

Would be great to also link to https://docs.cilium.io/en/v1.9/gettingstarted/ which is how users can get started. The existing link to https://docs.cilium.io/en/latest/bpf/ covers the low-level eBPF reference guide.

**Review by tgraf — CHANGES_REQUESTED** [review]

File: website/content/usage/_index.md
Comment (onmyflow) [reply website/content/usage/_index.md:172]: Great Ideas! Would you mind creating a suggestion as to how you would like it to be written? I'm not very familiar with Cilium, so It'll be great if somebody with more expertise could write something about the inner workings of it.

Comment (joestringer) [reply website/content/usage/_index.md:172]: Yeah sure :-) Looking again at the context and also @tgraf's feedback, maybe something like the following would suffice..?

```suggestion
[Cilium](https://cilium.io) is a CNI that provides eBPF-based Networking, Observability, and Security. As part of its kube-proxy replacement implementation, Cilium is able to preserve the source IP when performing load balancing decisions across the cluster through [Direct Server Return (DSR)](https://cilium.io/blog/2020/02/18/cilium-17#kube-proxy-replacement-with-direct-server-return). For more information and guides to deploy Cilium, see the [Cilium Getting Started Guides](https://docs.cilium.io/en/stable/gettingstarted/).
```

File: website/content/usage/_index.md
Comment (onmyflow) [reply website/content/usage/_index.md]: I've commited the suggestion of @joestringer to this PR. Feel free to add a suggestion aswell :)

**Comment by russellb** [comment]
The review from @champtar makes it sound like we shouldn't merge this as it stands, so I'm going to close it out now.  Feel free to reopen if you'd like to discuss further or propose additional changes.  Thanks!

**Comment by BloodyIron** [comment]
The MetalLB documentation sorely needs SourceIP solutions in its documentation. I've been spending over 3 years now trying to get a solution working between Calico and Cilium with RKE1/2 nodes via Rancher. I am currently working on testing RKE2+Calico for the SourceIP issue (Layer 2 ARP mode btw for MetalLB). So I highly recommend this topic be re-opened, even if the outcome needs to change in specific details.

The functional need for SourceIP retention hasn't gone away, and is sorely needed in the MetalLB documentation.

**Comment by BloodyIron** [comment]
@russellb tagging you as you were the one to close this.

**Comment by rata** [comment]
@BloodyIron using it at the k8s layer is not enough for you? I mean setting the [externTrafficPolicy](https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/#preserving-the-client-source-ip) in the service

**Comment by BloodyIron** [comment]
I have no idea how that relates to improving the documentation, and so far as I can tell no it is not relevant to what I'm talking about. I really don't see where you're coming from here.

**Comment by rata** [comment]
@BloodyIron where I'm coming from? What do you mean?

The question is if that would help (it always did the trick for me with metallb +BGP in bare metal clusters, it is probably accurate but I haven't thought if it works for layer 2 too, probably it doesn't). The first step to write documentation, is to know what works and that is why I'm asking.

If it doesn't work for you, can you please share a little more? Do you know why not or what happens? What does work for you?

**Comment by BloodyIron** [comment]
> @BloodyIron where I'm coming from? What do you mean?
> 
> The question is if that would help (it always did the trick for me with metallb +BGP in bare metal clusters, it is probably accurate but I haven't thought if it works for layer 2 too, probably it doesn't). The first step to write documentation, is to know what works and that is why I'm asking.
> 
> If it doesn't work for you, can you please share a little more? Do you know why not or what happens? What does work for you?

Well I don't know what you mean by "using it at the k8s layer"... using  _what_ and I don't know what you mean by "k8s layer"??? And externalTrafficPolicy, generally I have that set to Cluster for NGINX Ingress, and still not following what you mean by that part anyways.

It's not like I disagree with quality documentation, it's just I am nowhere near on the same wavelength as you... so I don't understand your questions thus-far to me.

**Comment by rata** [comment]
I mean setting the [externTrafficPolicy](https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/#preserving-the-client-source-ip) in the service to `Local`. That way, the service should see the client's source IP (iow, what you expect, the real source IP).

**Comment by BloodyIron** [comment]
> I mean setting the [externTrafficPolicy](https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/#preserving-the-client-source-ip) in the service to `Local`. That way, the service should see the client's source IP (iow, what you expect, the real source IP).

I don't want that. I want the traffic routed within the cluster as I don't want _every node_ ingesting traffic. No thank you.

**Pull request closed without merging**
